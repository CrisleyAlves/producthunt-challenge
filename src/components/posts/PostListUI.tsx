import React from "react";
import { faHeart } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import Avatar from "../avatar/Avatar";

type Post = {
  name: string;
  tagline: string;
  votesCount: number;
  url: string;
  thumbnail?: {
    url: string;
  };
};

type PostListUIProps = {
  posts?: {
    cursor: string;
    node: Post;
  }[];
};

export default function PostListUI({ posts = [] }: PostListUIProps) {
  return (
    <ul className="post__list">
      {posts.map(({ cursor, node: post }, index) => {
        return (
          <li
            className="post__list__item card card--rounded"
            key={`${cursor}-${index}`}
          >
            <div className="post">
              <Avatar
                classNames="avatar--rounded"
                avatarUrl={post?.thumbnail?.url || ""}
                altText="post iamge"
                ariaDescription="post iamge"
              />

              <div className="post__content">
                <h2 className="post__title">{post.name}</h2>
                <p className="post__description">{post.tagline}</p>
                <a href={post.url} target="_blank" className="post__link">
                  visit
                </a>
              </div>

              <div className="post__count">
                <FontAwesomeIcon className="icon" icon={faHeart} />
                <p className="post__count__value">{post.votesCount}</p>
              </div>
            </div>
          </li>
        );
      })}
    </ul>
  );
}
