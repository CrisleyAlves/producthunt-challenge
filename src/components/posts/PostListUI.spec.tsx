import React from "react";
import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import PostListUI from "./PostListUI";

const MOCKED_POSTS = [
  {
    cursor: "cursor1",
    node: {
      name: "Test Post 1",
      tagline: "First test post",
      votesCount: 42,
      url: "https://example.com/1",
      thumbnail: { url: "https://example.com/img1.png" },
    },
  },
  {
    cursor: "cursor2",
    node: {
      name: "Test Post 2",
      tagline: "Second test post",
      votesCount: 99,
      url: "https://example.com/2",
      thumbnail: { url: "https://example.com/img2.png" },
    },
  },
];

describe("PostListUI", () => {
  it("renders a list of posts", () => {
    render(<PostListUI posts={MOCKED_POSTS} />);
    expect(screen.getByText("Test Post 1")).toBeDefined();
    expect(screen.getByText("Test Post 2")).toBeDefined();
    expect(screen.getByText("First test post")).toBeDefined();
    expect(screen.getByText("Second test post")).toBeDefined();
    expect(screen.getAllByText("visit")).toHaveLength(2);
    expect(screen.getByText("42")).toBeDefined();
    expect(screen.getByText("99")).toBeDefined();
  });

  it("renders empty list when posts is empty", () => {
    const { container } = render(<PostListUI posts={[]} />);
    expect(container.querySelectorAll("li").length).toBe(0);
  });

  it("renders without crashing when posts prop is undefined", () => {
    const { container } = render(<PostListUI />);
    expect(container.querySelectorAll("li").length).toBe(0);
  });
});
