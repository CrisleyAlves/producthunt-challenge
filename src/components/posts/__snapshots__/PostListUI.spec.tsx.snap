// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostListUI > renders a list of posts 1`] = `
<div>
  <ul
    class="post__list"
  >
    <li
      class="post__list__item card card--rounded"
    >
      <div
        class="post"
      >
        <img
          alt="post iamge"
          aria-description="post iamge"
          class="avatar avatar--rounded"
          src="https://example.com/img1.png"
        />
        <div
          class="post__content"
        >
          <h2
            class="post__title"
          >
            Test Post 1
          </h2>
          <p
            class="post__description"
          >
            First test post
          </p>
          <a
            class="post__link"
            href="https://example.com/1"
            target="_blank"
          >
            visit
          </a>
        </div>
        <div
          class="post__count"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart icon"
            data-icon="heart"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
              fill="currentColor"
            />
          </svg>
          <p
            class="post__count__value"
          >
            42
          </p>
        </div>
      </div>
    </li>
    <li
      class="post__list__item card card--rounded"
    >
      <div
        class="post"
      >
        <img
          alt="post iamge"
          aria-description="post iamge"
          class="avatar avatar--rounded"
          src="https://example.com/img2.png"
        />
        <div
          class="post__content"
        >
          <h2
            class="post__title"
          >
            Test Post 2
          </h2>
          <p
            class="post__description"
          >
            Second test post
          </p>
          <a
            class="post__link"
            href="https://example.com/2"
            target="_blank"
          >
            visit
          </a>
        </div>
        <div
          class="post__count"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-heart icon"
            data-icon="heart"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
              fill="currentColor"
            />
          </svg>
          <p
            class="post__count__value"
          >
            99
          </p>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
